lib/charms/mongodb/v0/shards_interface.py

```diff
@@ -116,9 +141,27 @@ def _on_relation_event(self, event):
         try:
             logger.info("Adding shards not present in cluster.")
             self.add_shards(departed_relation_id)
-            # TODO Future PR, enable updating shards by listening for relation changed events
-            # TODO Future PR, enable shard drainage by listening for relation departed events
-        except PyMongoError as e:
+            self.remove_shards(departed_relation_id)
+        except NotDrainedError:
+            # it is necessary to removeShard multiple times for the shard to be removed.
+            logger.info(
+                "Shard is still present in the cluster after removal, will defer and remove again."
+            )
+            event.defer()
+            return
+        except OperationFailure as e:
+            if e.code == 20:
```

Comment:
Could we use constant instead of a magic number? 😄

Reply:
Of course, thank you!